### PR TITLE
common-lisp: add sly backend

### PR DESCRIPTION
**Pull request by doyougnu**
### Summary of Changes
Closes out #43. The only public facing changes are the addition of `symex-common-lisp-backend`. This defaults to `slime` but users can optionally set it to `sly`.

I tried to do some fancy things to reduce code duplication and the nested checks on this variable, but it didn't fit the project style. I don't think it will be a problem since these runtime checks should be `jit`d away.

I didn't see a place in the docs that says `symex supports the following lisps: ...` so I was unsure where to state:
```
Symex can use Sly as a backend for Common Lisp as opposed to Slime. To do so set symex-common-lisp-backend, 
for example you might have the following use-package call (for Doom emacs):

(use-package! symex
  :config
  (symex-initialize)
  (setq symex-common-lisp-backend 'sly)

  (map! :after evil
        :map evil-motion-state-map
        "\\" #'symex-mode-interface)

  :custom
  (symex-modal-backend 'evil))
```

### Public Domain Dedication

- [x] In contributing, I relinquish any copyright claims on my contribution and freely release it into the public domain in the simple hope that it will provide value.

(**Why**: The freely released, copyright-free work in this repository represents an investment in a better way of doing things called _attribution-based economics_. Attribution-based economics is based on the simple idea that we gain more by giving more, not by holding on to things that, truly, we could only create because we, in our turn, received from others. As it turns out, an economic system based on attribution -- where those who give more are more empowered -- is significantly more efficient than capitalism while also being stable and fair (_unlike_ capitalism, on both counts), giving it transformative power to elevate the human condition and address the problems that face us today along with a host of others that have been intractable since the beginning. You can help make this a reality by releasing your work in the same way -- freely into the public domain in the simple hope of providing value. Learn more about attribution-based economics at [drym.org](https://drym.org), tell your friends, do your part.)

### CODE REVIEW

File: symex-custom.el [hunk 1]
```diff
@@ -60,6 +60,10 @@
   :type 'list
   :group 'symex)
 
+(defcustom symex-common-lisp-backend 'slime
+  "Backend provider for Common Lisp interactive features; defaults to SLIME."
+  :type 'symbol
+  :group 'symex)
```
Comment (countvajhula) [inline symex-custom.el:66]: nit: Would be good to mention the possible options here, slime or sly. [Here's an example docstring](https://github.com/greghendershott/racket-mode/blob/52e993f3b39416eeb4df6262491d4a42d0b35232/racket-custom.el#L210-L225) that we could follow. I realize this also affects the evil option above :)

**Comment by countvajhula** [comment]
> I tried to do some fancy things to reduce code duplication and the nested checks on this variable, but it didn't fit the project style. I don't think it will be a problem since these runtime checks should be jitd away.

That's a good thought. It may be worth looking into something like this a little more as part of addressing #26  .

> I didn't see a place in the docs that says symex supports the following lisps: ... so I was unsure where to state:

```
(map! :after evil
        :map evil-motion-state-map
        "\\" #'symex-mode-interface)
```
Do you have symex mode mapped in place of the evil leader key `\`? Interesting UX!

You could put this in a top-level heading between "Usage and Customization" and "Tips" in the README, called something like "Lisp Flavors" and under a "Common Lisp" subheader (can leave out other Lisp flavors for now, and I can flesh them out later if necessary). Re: mapping the leader `\` to symex-mode-interface, that probably doesn't belong in the same section of the README since it's more of a suggested UX than a config for Common Lisp specifically. But if you are a fan of this approach and would like to suggest it to other users, then this could fall under the "Tips" section, along with some motivation for why it's nice.

**Comment by doyougnu** [comment]
> Do you have symex mode mapped in place of the evil leader key \? Interesting UX!

Ah I'm using [Doom Emacs](https://github.com/hlissner/doom-emacs) so `evil-leader` is bound  `SPC` and my `localleader` key is bound to `,`. This clears up `\` for symex! This way I have a single keybind to call `symex-mode-interface` and if I need to back out I just type my `evil-escape-key-sequence` (which is `fd`) and I'm back in evil normal mode. 

> Re: mapping the leader \ to symex-mode-interface, that probably doesn't belong in the same section of the README since it's more of a suggested UX than a config for Common Lisp specifically.

I totally agree here. I just got lazy and copy/pasted my config for doom :)

File: symex-custom.el (hunk 1 above)
Comment (doyougnu) [reply symex-custom.el:69]: See https://github.com/countvajhula/symex.el/pull/44/commits/8982040871c9658e231a2b678358e305da34c6a8

**Comment by countvajhula** [comment]
This looks great, thank you for your contribution!

**Pull request merged**